src/backend/gl/src/window/glutin.rs

```diff
@@ -43,60 +43,152 @@
 //! }
 //! ```
 
-use hal::{self, format as f, image};
+use hal::{self, format as f, image, memory as m};
+use hal::format::AsFormat;
 
-use {Backend as B, Device, PhysicalDevice, QueueFamily, Starc};
+use {
+    Backend as B, Device, PhysicalDevice, QueueFamily, Starc, Wstarc,
+    InstanceContext, Instance, Contexts, native, gl
+};
 
 use glutin::{self, GlContext};
 
+#[cfg(all(unix, not(target_os = "android")))]
+use glutin::wplatform::{XNotSupported, XConnection};
+#[cfg(all(unix, not(target_os = "android")))]
+use glutin::os::unix::{EventsLoopExt, WindowExt};
 
-fn get_window_extent(window: &glutin::GlWindow) -> image::Extent {
-    let px = window.get_inner_size().unwrap().to_physical(window.get_hidpi_factor());
-    image::Extent {
-        width: px.width as image::Size,
-        height: px.height as image::Size,
-        depth: 1,
+#[cfg(windows)]
+use glutin::os::windows::WindowExt;
+
+use libc;
+#[cfg(all(unix, not(target_os = "android")))]
+use x11_dl;
+
+use std::env;
+use std::sync::{Arc, Mutex};
+use std::os::raw::c_void;
+
+const BACKEND_PREFERENCE_ENV_VAR: &str = "GFX_OPENGL_UNIX_BACKEND";
+
+pub(crate) fn physical_to_extent(ex: &glutin::dpi::PhysicalSize) -> hal::window::Extent2D {
+    hal::window::Extent2D {
+        width: ex.width as _,
+        height: ex.height as _,
     }
 }
 
 pub struct Swapchain {
     // Underlying window, required for presentation
-    pub(crate) window: Starc<glutin::GlWindow>,
+    pub(crate) surface: Arc<SurfaceSwapchained>,
+    pub(crate) images: Vec<SwapchainImage>,
+
+    // The images' extents
+    pub(crate) extent: hal::window::Extent2D,
+}
+
+pub struct SwapchainImage {
+    // The fbo on the window's context
+    pub(crate) fbo: u32,
+
+    // The image we render too.
+    pub(crate) image: native::Image,
+    pub(crate) image_mem: native::Memory,
+
+    pub(crate) currently_acquired: Mutex<bool>,
 }
 
 impl hal::Swapchain<B> for Swapchain {
     fn acquire_image(
-        &mut self, _timeout_ns: u64, _sync: hal::FrameSync<B>
+        &mut self, timeout_ns: u64, _sync: hal::FrameSync<B>
     ) -> Result<hal::SwapImageIndex, hal::AcquireError> {
-        // TODO: sync
-        Ok(0)
+        // NOTE: If the extents don't match, it is *only* suboptimal. HAL doesn't
+        // currently distinguish between suboptimal and out of date, so we just
+        // have to return out of date. This is a note to whoever adds the ability to
+        // distinquish between the two.
+        let extent = physical_to_extent(
+            &self.surface
+            .surface_context
+            .window
+            .get_inner_size()
+            .unwrap()
+            .to_physical(self.surface.surface_context.window.get_hidpi_factor())
+        );
+
+        if extent != self.extent {
+            return Err(hal::AcquireError::OutOfDate);
+        }
+
+        use std::time::Instant;
```

Comment:
Respecting the timeout could be split into a separate PR, if someone would like.